### DISCUSSION TITLE
Support sourcemaps for client-only code?

### DISCUSSION THREAD

**Discussion by kentcdodds** [remix-run/remix#5202] [Proposals]
Due to a security issue, sourcemaps were made opt-in: https://github.com/remix-run/remix/pull/350

Luckily for me all my code is open source anyway so I can just enable them. But I think there's probably a lot of value for people who don't want their server-side source to be visible but want client-side sourcemaps. What are the chances we could get an option (or a default) that adds sourcemaps but only for the client-side code?

I'm not even sure how possible that would be, but if it were possible to have sourcemaps for my modules that somehow magically doesn't include any server code that would be a stellar default I think.

**Comment by jacob-ebey** [comment]
I will definitely be looking into how we can optimize to allow sourcemaps in the browser without the server portions being included. It's strange though because at that point it's not a true sourcemap, but more of a partial-sourcemap.

**Comment by kiliman** [comment]
The compiler makes 2 passes. One for the server build and the other for client. Each one includes the `sourcemap: true/false` flag, so it should be possible to generate just for client code. Just need to add a separate option for client/server sourcemaps.

from `createBrowserBuild()`
https://github.com/remix-run/remix/blob/e6e6e1cd6ad545aaf39c7780d58f3293fdfe8682/packages/remix-dev/compiler.ts#L300

from `createServerBuild()`
https://github.com/remix-run/remix/blob/e6e6e1cd6ad545aaf39c7780d58f3293fdfe8682/packages/remix-dev/compiler.ts#L339

**Comment by sergiodxa** [comment]
But the sourcemap is not going to still show the loader/action code of each route? Because that's the source code used to create both the server and browser build.

**Comment by kiliman** [comment]
Ah, I see what you mean. Yeah, I just checked the sourcemap generated for the route and "sourcesContent" includes the entire source of the file, not just the stripped/client version. I supposed you could post-process the sourcemap and simply hide all the code that's not supposed to be on the client (replace each line with a blank to preserve line numbers). You'd only need to do that on the production build.

**Comment by kiliman** [comment]
Hmm. Why is this closed? It’s still an issue although there is a warning now that you probably shouldn’t expose your source map on production. 

I want to fix this so it post-processes the map file and strips server code from the route.

**Comment by machour** [comment]
Converting this to a Proposal discussion, as per our [Development Process](https://remix.run/docs/en/v1/pages/contributing#development-process).

**Comment by ariofrio** [comment]
What I actually would want, I think, is the server sourcemap and the client sourcemap to be in different files, so that I can have the server sourcemap available for Node stacktraces in logs but block it from being accessible by browsers, while client sourcemaps are accessible by browsers to help with debugging.

**Reply by sergiodxa** [reply]
The server source maps are placed together with the server code, e.g. `build/index.js.map`, and the client source maps are in the `public/build` folder. You could delete the `.map` files in `public/build` and keep the ones for the server.

**Reply by brettdh** [reply]
So is the warning presented with `remix build --sourcemap` still accurate? If the server and client code have separate source maps, is server code still exposed if source maps in `public/build` are published? (Setting aside the code that runs on both client and server, that is.)

**Reply by kiliman** [reply]
Yes, unfortunately even though Remix creates separate bundles for server and client, the sourcemap still contains the entire source of your route module (including loaders, etc.). This happens with both the default compiler as well as Vite.

The sourcemap will need to be post-processed to strip out the server code.

I'm using Sentry, so I upload the raw sourcemap to Sentry and then delete them from the `public/build` folder during the deployment process. So any errors on the client will show a minified stacktrace, but the Sentry log will show the correct source line numbers.

**Reply by brettdh** [reply]
@kiliman thanks for the quick response. For our use, I think we'll probably just accept the exposure of this source code for now, as our Remix server code is primarily [backend-for-frontend](https://remix.run/docs/en/main/guides/bff), calling a separate API written in Python.

We're also looking at using Sentry, but also having the ability to debug our Remix app running in AWS using the browser devtools (source maps included) is pretty important to us. 

> The sourcemap will need to be post-processed to strip out the server code.

How much effort would this be? I may be able to take a crack at it if you can point me in the right direction.

**Reply by kentcdodds** [reply]
I'm not sure how much work that would be, but I can tell you that many people would benefit greatly from that effort. Far beyond the Remix community. Many communities are going to have similar challenges as we see many of them write server and client code in the same file.

**Reply by kiliman** [reply]
I did an experiment and it seems that as long as you replace any "server" code with a blank line, the rest of the sourcemap works properly (line/column) references.

### before processing
<img width="588" alt="image" src="https://github.com/remix-run/remix/assets/47168/91e40f9e-528a-4c65-80dd-fae4c700b560">

### after removing server code from source map
I removed the server imports and the loader function

<img width="589" alt="image" src="https://github.com/remix-run/remix/assets/47168/1aa1b711-1911-4399-94cb-5d6e97f3df52">

So basically we need a way to determine which code is server-side only and replace it with blank lines.

**Reply by kiliman** [reply]
I went ahead and created a CLI to strip sourcemaps.

You can try it out here:

https://github.com/kiliman/strip-sourcemaps
```bash
USAGE: npx strip-sourcemaps [options]
This script strips server source code from the sourcemaps in the [build-path].

options:
-h, --help
  show help
-b, --build-path <path>
  path to the client build directory (default: ./public/build)
-o, --output <path>
  path to the stripped output directory (default: none)
  useful for verifying that the code is stripped correctly
-e, --exports <string> (default: loader,action)
  comma-separated list of exports to strip
-i, --imports <string> (default: *)
  comma-separated list of imports to strip (only supports * or empty string)
```

**Reply by kiliman** [reply]
It parses the AST and removes imports and exports (loaders/actions)

**Reply by brettdh** [reply]
Very cool, thanks. I will hopefully get a chance to try this out soon (this week).

**Reply by brettdh** [reply]
Got a chance to try this today. I got an error with this message:

```
  messageText: "Cannot find module 'remix-development-tools/index.css' or its corresponding type declarations.",
```

This was part of a large object printed after the list of source files found. It looks to me like this is the result of `parseFromFiles, [here](https://github.com/kiliman/strip-sourcemaps/blob/main/src/index.ts#L130).

The `text` field of the error contained this import statement:
```
      'import rdtStylesheet from "remix-development-tools/index.css"\n' +
```

which comes from my `app/root.tsx` file, as instructed here: https://github.com/Code-Forge-Net/Remix-Dev-Tools

The source code is proprietary, but please let me know if you want me to dig some other specific info out of the error.

**Reply by kiliman** [reply]
Yes, the current parser doesn't work well for our use case.

I've already rewritten it using the `ts-morph` package. Its AST handling is much better and easier.

I've changed my strategy from stripping "server code" to assuming we'll strip _everything_ and only include "client code" and what it references. This will be much safer. It will also strip out comments (which may include sensitive info). 

I hope to have something ready over the weekend.

**Reply by brettdh** [reply]
No rush, but since it's been a bit, I wanted to check in on this.

**Reply by kiliman** [reply]
It is still on my TODO list. I currently have family visiting this week. Also, I'm focusing on getting my Remix Packs project finished.

**Comment by mmachatschek** [comment]
For anyone following: https://github.com/remix-run/remix/pull/8613

**Reply by brettdh** [reply]
I'm trying to understand what, if anything, has changed with regard to server code appearing in client sourcemaps in the Vite plugin. I'd made a note to come back to this after upgrading to a Remix version with Vite support, thinking that there would be some change, but looking back at @kiliman 's earlier comment, perhaps not:

> Yes, unfortunately even though Remix creates separate bundles for server and client, the sourcemap still contains the entire source of your route module (including loaders, etc.). This happens with both the default compiler as well as Vite.
> 
> The sourcemap will need to be post-processed to strip out the server code.
